### DISCUSSION TITLE
Common Approach for Chat Memory on Workflow

### DISCUSSION THREAD

**Discussion by daryadifoo** [langgenius/dify#4058] [General]
I want to know why Chat Memory only stores the first query and last answer of the entire workflow? This approach makes chat history inappropriate when we use more than 1 llm. Let's say we use 3 llm nodes in 1 workflow. So llm 2 will receive the history from the query in llm 1 and the answer from llm 3. Meanwhile llm 2 should receive the history from llm 1's answer and answer from llm 2 itself.

**Comment by daryadifoo** [comment]
Another example:
![Untitled-2024-05-03-1330](https://github.com/langgenius/dify/assets/38455104/7afbe74d-e948-4f44-9790-fb6e0c45b7a7)

**Reply by xiaobinwu24** [reply]
I have encountered this issue as well.

**Comment by skywolf123** [comment]
Yes, the design of this chat memory is very strange, taking the whole workflow as a whole or an agent as a whole, and storing the input and output of this whole.

It is completely different from the chat history we usually understand, that is, the history function used by chat clients such as lobe.  

Caused us to be unable to automate the chat flow with dify.

**Comment by Yevanchen** [comment]
Can the read and write of conversation - variables be used to solve this problem?

**Reply by skywolf123** [reply]
> Can the read and write of conversation - variables be used to solve this problem?

It can't be solved, because the prompt part of the LLM node (i.e., the input) cannot be obtained, only the output of the LLM node can be used when the variable is assigned

**Comment by Yevanchen** [comment]
> I want to know why Chat Memory only stores the first query and last answer of the entire workflow? This approach makes chat history inappropriate when we use more than 1 llm. Let's say we use 3 llm nodes in 1 workflow. So llm 2 will receive the history from the query in llm 1 and the answer from llm 3. Meanwhile llm 2 should receive the history from llm 1's answer and answer from llm 2 itself.

the chatflow in dify only records the QA pairs in the chat window, which already covers scenarios of chat history. I completely understand that "the history from llm 1's answer and answer from llm 2 itself" constitutes a complete context for the llm2 node. Could you provide more detailed scenarios to help me further understand the purpose behind your ideas? @skywolf123 @daryadifoo If there’s a delay in response, my email is evanchen@dify.ai.Thank you for your suggestions and feedback again
